src/odemis/driver/tfsbc.py

```diff
@@ -0,0 +1,376 @@
+            rq = self._serial.write_registers(BEAMDEFL_LX, values, unit=SLAVE_UNIT)
+        except IOError:
+            self._reconnect()
+            raise IOError("Failed to write registers of beam control firmware, "
+                          "restarted serial connection.")
+
+    def _read_registers(self):
+        """
+        Read all four registers. Try to reconnect to device in case connection was lost.
+        :return (list of 4 ints): register values (-x, -y, x, y)
+        """
+        try:
+            # write all registers together (starting at lower x register (=0x01))
+            rr = self._serial.read_holding_registers(BEAMDEFL_LX, 4, unit=SLAVE_UNIT)
+            return rr.registers
+        except IOError:
+            self._reconnect()
+            raise IOError("Failed to write registers of beam control firmware, "
+                          "restarted serial connection.")
+
+    def _reconnect(self):
+        """
+        Attempt to reconnect. It will block until this happens.
+        On return, the hardware should be ready to use as before.
+        """
+        num_it = 5
+        self.state._set_value(model.HwError("Beam deflection controller disconnected"), force_write=True)
+        logging.warning("Failed to write registers, trying to reconnect...")
+        for i in range(num_it):
+            try:
+                self._serial.close()
+                self._serial = None
+                self._port = self._findDevice(self._portpattern, self._serialnum)
+                self._serial = self._openSerialPort(self._port)
+                logging.info("Recovered device.")
+                break
+            except IOError:
+                continue
+        else:
+            raise IOError("Failed to reconnect to beam deflection controller.")
+        self.state._set_value(model.ST_RUNNING, force_write=True)
+
+    def updateMetadata(self, md):
+        if model.MD_CALIB_BEAMSHIFT in md:
+            # Check format
+            bs = md[model.MD_CALIB_BEAMSHIFT]
+            try:
+                if not len(bs) == 4:  # 4 tuples required
+                    raise ValueError("Invalid MD_CALIB_BEAMSHIFT metadata %s: 4 tuples required." % (bs,))
+                if not all(len(val) == 2 for val in bs):  # each of the 4 values is a tuple of 2
+                    raise ValueError("Invalid MD_CALIB_BEAMSHIFT metadata %s: Two values per tuple required." % (bs,))
+                if not all(all([type(val) in (int, float) for val in tup]) for tup in bs):  # each element is a number
+                    raise ValueError("Invalid MD_CALIB_BEAMSHIFT metadata %s: Values must be numbers." % (bs,))
+            except Exception as ex:
+                raise ValueError("Invalid MD_CALIB_BEAMSHIFT metadata %s, ex: %s" % (bs, ex,))
+
+            # If MD_CALIB_BEAMSHIFT was already available, transform current shift to fit new metadata value.
+            if self._metadata.get(model.MD_CALIB_BEAMSHIFT):
```

Comment:
You could write more explicitly `if model.MD_CALIB in self._metadata`.
However, I don't understand why doing it only if there was already some calibration. On the first time that the component gets a calibration, the hardware already has some values, so there is already a shift => directly use `md[model.MD_CALIB]`.